### ISSUE TITLE
Check the value of an ArrayFire array in Visual Studio or CLion

### ISSUE THREAD

**Issue by wuyuanyi135** [arrayfire/arrayfire#2916]
Hello,

I wonder whether the debugger display of the array value has been implemented. On Windows, either CLion (with VS compiler and LLDB) or Visual Studio, the variable is shown with only one field "arr". Is there any recommended practice to visualize the variable during breakpoint debugging?

Related issue:
* #268

**Comment by umar456** [comment]
We haven't created a debugger view for the ArrayFire array class for Visual Studio. It is something that would be super helpful for everyone and would be happy to accept a PR with this change. Here is some resource on how to do that in Visual Studio:

https://docs.microsoft.com/en-us/visualstudio/debugger/create-custom-views-of-native-objects?view=vs-2019

**Comment by wuyuanyi135** [comment]
Hmm. Seems showing the whole content of a large array is considerably hard without tweaking the debugger.

Is it possible to have an easy way to display some information such as array dimension, dtype, etc? Right now the debugger shows nothing but a head pointer to the internal arr data, and using the `af_` methods to retrieve those information is nontrivial 
* some requires output variable
* most unused symbols are not linked, so they are not available during debugging.

In terms of the Array visuallization, is it more feasible to have behavior like the `__str__` method like of the python `numpy` library? More specifically, attachingan optional variable containing the string like `AF_PRINT` to an Array object?

**Comment by wuyuanyi135** [comment]
> Additional Info:
> 
> * Natvis files are detected by cmake when added as sources for only Visual Studio generators only.
> * The above support is only available from 3.7 based on the [tags for the relevant commit](https://gitlab.kitware.com/cmake/cmake/-/commit/a6290cffaec49fc06875d6f0379af7d712b90bb3) in cmake project.
> 
> Having said that, I don't think view of `af::array` type can be achieved using a simple natvis file alone since the that data actually resides on GPU memory unlike other custom types.

Agreed. Displaying the variables on GPU memory needs some magic to automatically retrieve to the host. But if the CPU works well I think it is possible to switch between them for debugging and releasing, if no CUDA specific code involved.

**Comment by wuyuanyi135** [comment]
@umar456 In addition, I saw the Windows tag for this issue. On Linux or GCC, Are the debugging features for ArrayFire provided?

**Comment by umar456** [comment]
No. I guess not. You will need to add a custom gdb and lldb pretty printers to add support for debuggers.

Here is the resource here:
https://sourceware.org/gdb/onlinedocs/gdb/Writing-a-Pretty_002dPrinter.html

**Comment by 9prady9** [comment]
> Is it possible to have an easy way to display some information such as array dimension, dtype, etc? Right now the debugger shows nothing but a head pointer to the internal arr data, and using the `af_` methods to retrieve those information is nontrivial

dim4 type should be readily available even now. But yes, the meta data associated to af_array can't be fetched without somehow using C API via the debugger framework. I am only aware of natvis approach.

> In terms of the Array content, is it more feasible to have behavior like the `__str__` method like of the python `numpy` library? More specifically, attachingan optional variable containing the string like `AF_PRINT` to an Array object?

I am not sure how that is potentially possible. @umar456 any thoughts ?

**Comment by umar456** [comment]
> > In terms of the Array content, is it more feasible to have behavior like the `__str__` method like of the python `numpy` library? More specifically, attachingan optional variable containing the string like `AF_PRINT` to an Array object?
> 
> I am not sure how that is potentially possible. @umar456 any thoughts ?

There isn't an efficient way to do this for the GPU on most current platforms. It may be possible sometime in the future once we have coherent accesses to GPU memory but it may not be as efficient as not having this feature. I think pretty printers and custom viewers are the best way to handle this.
